### ISSUE TITLE
Override content part field editor view

### ISSUE THREAD

**Issue by MikeKry** [OrchardCMS/OrchardCore#11046]
Problem: Custom views or Admin templates can't be applied to fields of parts, even though these should work as they are defined in documentation ([PartType]\_\_[FieldName], [PartType]\_\_[FieldName]\_\_[ShapeType])

Reproduction:
I have created text field "test" inside standard "BagPart"
![image](https://user-images.githubusercontent.com/52829889/150409706-da55612a-0b96-43f6-b3e3-324fdbc8ee5d.png)

I would like to override default EDIT view of this field using Admin templates (or custom views). 
I have tried every possible combination of template names from documentation: https://docs.orchardcore.net/en/dev/docs/reference/modules/Templates/#fieldtype__shapetype.
E.g. "BagPart__test", "BagPart_Edit__Test", "BagPart_Edit__Test_Edit" ,...

![image](https://user-images.githubusercontent.com/52829889/150409755-87a5c739-15c4-4dd4-a370-573f0a0ac712.png)

Only "TextField_Edit" works, but that is not related to part.

for placements these differentiators do work, but for views I could not figure out if it is even working or I am doing something wrong? 

2/ I could not find if there is possibility to completely disable automatic rendering of fields from content parts, so dev would need to create view for whole content part including fields? in this case, field "test" would be in "BagPart.Edit.cshtml" instead of custom view

Thanks for reply.

**Comment by MikeKry** [comment]
Only solution I have found currently is to create a placement for this field with alternate definition, 
eg.
`[
  {
    "place": "Content:0",
    "displayType": null,
    "differentiator": "BagPart-test",
    "alternates": ["CustomAlternate"],
    "wrappers": null,
    "shape": null
  }
]`

**Comment by MikeKry** [comment]
Might be related to https://github.com/OrchardCMS/OrchardCore/issues/9174
but BagPart is not dynamic part.

**Comment by MikeKry** [comment]
I have also tried to debug whole differentiator and alternate adding process in ContentfieldDisplayDriver, but it seems that really only "TextField_Edit" is being added. 

I have found a possible way to override it in code using IShapeTableProvider, so I will try that..

**Comment by ns8482e** [comment]
@MikeKry i guess the editor shape can be defined via custom field  editor https://docs.orchardcore.net/en/latest/docs/reference/modules/ContentFields/#creating-custom-editors

**Comment by MikeKry** [comment]
@ns8482e 

This may be an option in certain cases, but I need to change editor in relation with content part. If I would create custom editor, it would be available anywhere in CMS which may not be wanted (certainly not in my case).

**Comment by Skrypt** [comment]
This could work. Though, need to see if he can differentiate from a BagPart editor and standard editor in that shape template.

**Comment by MikeKry** [comment]
@jtkech 

I mean, if I create custom editor for TextField e.g. TextField_Edit__CustomEditor, then I can choose this editor anywhere - in part, content type etc. but I need to customize editor only for BagPart in relation with TextField and do not allow to use this customized editor anywhere else.

I thought that should be addressed by template [PartType]\_\_[FieldName]\_\_[ShapeType] => BagPart_TextFieldName_CustomEditor

**Comment by Skrypt** [comment]
I don't think Edit and Create shape templates have alternates. The only thing you can override is each Field editor template as of now.

**Comment by MikeKry** [comment]
@Skrypt 

uh, okay, that at least explains my issue :) I guess it can't be added somehow easily?


(might be good to emphasize this fact in Templates documentation, I guess I wont be the only one struggling with this one)

**Comment by Skrypt** [comment]
This is indeed a common request as we have a lot of different issues opened about it and no PR yet submitted.
Adding alternates should be relatively simple.

Duplicate of #6784
https://github.com/OrchardCMS/OrchardCore/discussions/11044#discussioncomment-2008815

**Comment by jtkech** [comment]
@MikeKry Okay I got it

Yes we use custom editor modes intended to be used in the context of some content types
But yes nothing prevent to select them in the context of another type where it doesn't make sense
At least we take care about the name of the editor mode so that it is clearer where it can be used.

**Comment by MikeKry** [comment]
@Skrypt 

yes these issues are definitely related, but not completely same
#6784 - more about editing dynamic parts than fields of parts
#11044 - this one is one level deeper - changing fields of content type nested in bag part 

Anyway, I think there are 2 solutions:
1) add template rules for editor shapes (i think preferrable, because there already are templates for other display modes)
2) if there would be just something like migrations to setup alternates, it could be enough
e.g. AlterPartDefinition("BagPart), part => part.WithField("Text", field => field.AddAlternates("Alt1", "Alt2")....
or similliar in startup 

There already is something similliar for liquid "shape_add_alternates" but that is not useful in this context.

**Comment by Psichorex** [comment]
> @Skrypt
> 
> yes these issues are definitely related, but not completely same #6784 - more about editing dynamic parts than fields of parts #11044 - this one is one level deeper - changing fields of content type nested in bag part
> 
> Anyway, I think there are 2 solutions:
> 
> 1. add template rules for editor shapes (i think this could be prefered solution, because there already are templates for other display modes). Including custom editors of parts - "BagPart_Edit__CustomPartEditor__Field_Edit__CustomFieldEditor
> 2. if there would be just something like migrations to setup alternates, it could be enough
>    e.g. AlterPartDefinition("BagPart), part => part.WithField("Text", field => field.AddAlternates("Alt1", "Alt2")....
>    or similliar in startup
> 
> There already is something similliar for liquid "shape_add_alternates" but that is not useful in this context.

Isn't it possible to use IShapeProvider implementation to enable or disable certain shapes to render in an exact editor ?
For example as in #11044 (my issue) I am wondering that is that possible to prevent the FieldEditor shape to be rendered with IShapeProvider using something like  `.OnDisplaying(displaying => displaying.ChildContent = HtmlString.Empty);`  and somehow dig out a contentitemid to differentiate whether it is a BagPart creation or not?
If this is possible it may help OP with his issue

**Comment by ns8482e** [comment]
@Psichorex I can't find link to actual comment  (will share if find the link)- But - It's design decision that  - editor experience is based on editor mode defined in settings, hence there is no alternates registered  out-of-the-box for "Edit" for content fields based on ContentPart or ContentType. 

Each Editor mode  has it's own `ShapeType` meaning `TextField_Edit__Multiline` and `TextField_Edit` are not same shape type in other words placement of `TextField_Edit` doesn't apply to `TextField_Edit__Multiline`.

If you prefer not to implement custom edit modes for fields then you can provide your own shape using placement for standard editor based on part/type using placement

```json
"TextField_Edit" : {
    "shape" : "MyCustomEditor"
    "contentType": [ "MyType" ],
     "contentPart": [ "BagPart" ],
}
```

**Comment by Psichorex** [comment]
> @Psichorex I can't find link to actual comment (will share if find the link)- But - It's design decision that - editor experience is based on editor mode defined in settings, hence there is no alternates registered out-of-the-box for "Edit" for content fields based on ContentPart or ContentType.
> 
> Each Editor mode has it's own `ShapeType` meaning `TextField_Edit__Multiline` and `TextField_Edit` are not same shape type in other words placement of `TextField_Edit` doesn't apply to `TextField_Edit__Multiline`.
> 
> If you prefer not to implement custom edit modes for fields then you can provide your own shape using placement for standard editor based on part/type using placement
> 
> ```json
> "TextField_Edit" : {
>     "shape" : "MyCustomEditor"
>     "contentType": [ "MyType" ],
>      "contentPart": [ "BagPart" ],
> }
> ```

Can I even implement a custom edit cshtml for content type/part? The problem is that I should need a different edit page for my own Content Type when in a BagPart. If I implement placement as you suggested wouldn't it replace everything and after that it will always call that edit.cshtml when creating the item?

**Comment by ns8482e** [comment]
Wild guess and not tested but if you are trying `[PartType]__[FieldName]__[ShapeType]` here,  
PartType = 'BagPart'
FieldName = 'Test'
ShapeType = 'TextField_Edit' (standard editor)

then try this  `BagPart_Edit__Test__TextField_Edit` or (`BagPart.Test.TextField.Edit.Edit.cshtml`)  and see if that works?

**Comment by Psichorex** [comment]
> Wild guess and not tested but if you are trying `[PartType]__[FieldName]__[ShapeType]` here, PartType = 'BagPart' FieldName = 'Test' ShapeType = 'TextField_Edit' (standard editor)
> 
> then try this `BagPart_Edit__Test__TextField_Edit` or (`BagPart.Test.TextField.Edit.Edit.cshtml`) and see if that works?

Above mentioned that only fields can have overriden edits.
So here am I trying to figure out how could I make different edits for items in bagpart and out. JavaScript hiding is possible, but a not that nice solution. My main problem is still is that I can't find a grip on how to identify if a BuildRenderAsync for the editor shape is called from a BagPart or from the Create menu.

**Comment by MikeKry** [comment]
@ns8482e 

Using placements as you mention is indeed a solution probably for all three problems - but it can't be done programmatically, it has to be done either through admin UI or placements.json  ile, which is highly impractical. 

@Psichorex 

You can actually detect if you are in BagPart through IShapeTableProvider, I have tried that and it is an (ugly) alternative if this will not be solved. 

this is not the exact code as I have it on other computer, but you can try this:
` public class test : IShapeTableProvider
    {
        public void Discover(ShapeTableBuilder builder)
        {
            builder.Describe("TextField_Edit")
                .OnDisplaying(displaying =>
                {
                    var editor = displaying.Shape;

                    if ((editor as EditTextFieldViewModel).PartFieldDefinition.PartDefinition == "BagPart" && check desired name etc.)
                    {
                        editor.Metadata.Alternates.Add("TextField_Hidden"); // or anything else
                    }
                });
        }
    }`

**Comment by Psichorex** [comment]
> @ns8482e
> 
> Using placements as you mention is indeed a solution probably for all three problems - but it can't be done programmatically, it has to be done either through admin UI or placements.json file, which is highly impractical.
> 
> @Psichorex
> 
> You can actually detect if you are in BagPart through IShapeTableProvider, I have tried that and it is an (ugly) alternative if this will not be solved.
> 
> this is not the exact code as I have it on other computer, but you can try this:
> 
> ```
> public class test : IShapeTableProvider
>     {
>         public void Discover(ShapeTableBuilder builder)
>         {
>             builder.Describe("TextField_Edit")
>                 .OnDisplaying(displaying =>
>                 {
>                     var editor = displaying.Shape;
>                     if ((editor as EditTextFieldViewModel).PartFieldDefinition.PartDefinition == "BagPart" && check desired name etc.)
>                     {
>                         editor.Metadata.Alternates.Add("TextField_Hidden"); // or anything else
>                     }
>                 });
>         }
>     }
> ```

Is this method Discover() will run every time a shape is being rendered?

**Comment by MikeKry** [comment]
@Psichorex 

It does run every time for "TextField_Edit", but I think the same applies to placements, alternates and shapes anyway. I prefer to not use that solution though it could work.

**Comment by Psichorex** [comment]
@MikeKry Is TextField_Edit what is rendered when there is a ContentPickerField that is attached?

**Comment by ns8482e** [comment]
> Is this method Discover() will run every time a shape is being rendered?

`Discover()` runs only once ( and when features are enabled/disabled). However `OnDisplaying` runs every time just before shape is being displayed.

**Comment by MikeKry** [comment]
I think this issue should be transformed into some clear PR or something.

During some research I have found out that probably also editor options for parts are using templates? 
E.g. I have added "CustomEditor" to bagpart, created view "BagPart-CustomEditor.Edit.cshtml" but it still goes into "BagPart.Edit.cshtml"

**Comment by Skrypt** [comment]
Have you assigned the appropriate editor configuration on the field after adding the new shape?

**Comment by deanmarcussen** [comment]
Good to see you've discovered how to use both placements or `IShapeTableProvider`, both of these are recommended ways to affect placement / alternates.

The issue @Skrypt linked https://github.com/OrchardCMS/OrchardCore/issues/6784 is basically a duplicate of this, and in terms of any pr, remains the same.

i.e. add out of the box alternates to edit shapes, for parts and fields, in the base driver.

**Comment by MikeKry** [comment]
> Have you assigned the appropriate editor configuration on the field after adding the new shape?

Yes, I looked also in ContentPartDisplayDriverTPart.cs and it seems that it also does not have editor alternates, only for display. 

@deanmarcussen 

Ok , that gives sense, so I guess it's open for anyone who wants to add this feature?

**Comment by MikeKry** [comment]
Just to clarify: My statement that placements can't be added is false, it can be done through ShapeAlterationBuilder.

**Comment by deanmarcussen** [comment]
> Ok , that gives sense, so I guess it's open for anyone who wants to add this feature?

Yes, that's kinda how it works.

The alternates will be similar to what we do for on display, but without a display type, as there is no concept of a display type, when building an editor.

**Comment by MikeKry** [comment]
> I think this issue should be transformed into some clear PR or something.
> 
> During some research I have found out that probably also editor options for parts are not using templates? E.g. I have added "CustomEditor" editor option by creating "BagPart-CustomEditor.Options.cshtml" to bagpart, created view "BagPart-CustomEditor.Edit.cshtml" but it still goes into "BagPart.Edit.cshtml"

just FYI: This issue seems fixed in new versions, after updating Bagpart-CustomEditor.Edit.cshtml is called, was there any pr about this?

**Comment by ns8482e** [comment]
> just FYI: This issue seems fixed in new versions, after updating Bagpart-CustomEditor.Edit.cshtml is called, was there any pr about this?

@MikeKry @Skrypt  @sebastienros 

May be  PR #11098 has solved this
